lib/Dialect/TT/IR/TTOpsTypes.cpp

```diff
@@ -170,8 +170,75 @@ LayoutAttr LayoutAttr::get(
              collapseIntervals, oobVal);
 }
 
-llvm::SmallVector<int64_t> LayoutAttr::getStride() const {
-  return llvm::SmallVector<int64_t>();
+// From the logical shape of the tensor and the affine map of the layout,
+// compute the physical shape of the tensor, i.e the shape of the tensor
+// after the dimensions have been collapsed onto a grid.
+llvm::SmallVector<int64_t>
+LayoutAttr::getPhysicalShape(ArrayRef<int64_t> logicalShape) const {
+  llvm::SmallVector<int64_t> physicalShape(getGrid().getShape().size());
+  SmallVector<AffineExpr> logicalShapeExprs(
+      llvm::map_range(logicalShape, [context = getContext()](std::int64_t e) {
+        return getAffineConstantExpr(e - 1, context);
+      }));
+
+  for (size_t i = 0; i < physicalShape.size(); i++) {
+    AffineExpr expr = getLinear().getResult(i);
+    AffineExpr constantExpr = expr.replaceDims(logicalShapeExprs);
+    std::int64_t constant =
+        llvm::cast<AffineConstantExpr>(constantExpr).getValue() + 1;
+    physicalShape[i] = constant;
```

Comment:
This value needs to be aligned up to be a multiple of the respective grid dimension.

Reply:
Note the stride is 36 for inner dim, for logical shape 35 parallelized on 2 cores.

<img width="715" alt="Screenshot 2024-07-17 at 2 19 42 PM" src="https://github.com/user-attachments/assets/ee58a2cb-e79d-4c30-b432-4f7f6e51922c">

Reply:
Also, above diagram is a bit confusing, it calls "physical shape" something different from how you've defined your function here, but I think that's OK.  I'm actually not sure that physical tensor shape the way the tool has defined it is useful, we can redefine this in the document as a follow on, because I like your interpretation of physical shape better.

Reply:
I think we are misaligned on what "stride" should mean. I've taken the plain definition of stride as a way to map from logical space of tensor into the contiguous physical memory. In that context, main use for this would be that FE can allocate the memory in advance and know how to interpret it.

Once cores are taken into account and there is no contiguous memory, I'm not sure how to interpret stride? What will the context be in which that stride is used? Do we need a different terminology for this?

Reply:
I think we have the same definition of stride, but stride doesn't necessarily mean that the data must be tightly packed.  For example, you might have 30 scalars per row, but want every tensor row to start on an address multiple of 32, this would mean you need to program a stride of 32, not 30.

Stride here does do a bit of double duty, because it's constrained in such a way that it guarantees divisibility of the grid on the respective dimension.  This is how we can infer from how the shard shape should look in programming the memref.  I should write additional documentation for this point, but it is touched on in the padding section: https://tenstorrent.github.io/tt-mlir/specs/tensor-layout.html#padding

Reply:
Was thinking about this a bit more this afternoon.  All of my points are a bit moot until we have ttnn tensor layout support and the ability to do zero copy.  For now you could pretty much flexibly program stride however you want and it'll work because we have to bounce the tensor through ttnn cpu tensor anyway.

We can check this in as is for now and once ttnn layout is in place I'll file an issue for zero copy support and then we'll need the strides to be tightly coupled to what the compiler and device expect.